[WNMGDS-2171] CSS reset documentation (#2399)
CMSgov/design-system: packages/docs/content/migration-guides/introducing-a-css-reset.mdx

If you're curious to see what changes have been made, you can preview our reset styles in [Storybook](/storybook/?path=/story/reset-overview--overview). We'll break down the major changes for you to review below.

## Global defaults

The CSS reset introduces globally-defined defaults that will impact the entire application once the CSS. These defaults include (but aren't limited to):

- `box-sizing: border-box` is set on every element — including `*::before` and `*::after`. This ensures that the declared width of element is never exceeded due to padding or border.
- The `<html>` element has `scroll-behavior: smooth` set on `:focus-within`. This is a CSS-only solution that provides smooth anchor scrolling.
- Animations, transitions, and smooth scroll are removed on all elements - including `*::before` and `*::after` for users who have `prefers-reduced-motion` enabled to respect their preferences.
- The `<body>` element effectively replaces the `.ds-base` class, defining `font-family`, `font-size`, `line-height`, and `color`.
  - In addition, `<body>` now sets `background-color` to a theme's specified background.

> - The `outline` is removed on elements with the `tabindex="-1"` declared. This is removes focus rings for non-interactive elements, e.g., when activating a skip link will then send focus to a `<main>` or heading element.

Reviewer: ```suggestion
- The `outline` is removed on elements with the `tabindex="-1"` declared. This is removes focus rings for non-interactive elements. For example, when activating a skip link it will usually send focus to a `<main>` or heading element.
```
This might make it easier to read and simplifies the sentence structure

- To avoid [margin collapse](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing), all elements will only rely on their respective top margins to better control layout. All bottom margins have been removed.

## Typography

### Headings

All heading elements will receive the font styles defined by their parent element using the `font: inherit` rule. 

> [`font`](https://developer.mozilla.org/en-US/docs/Web/CSS/font) is a shorthand property that encompasses several font styles, but the key thing to note is a heading will now inherit the `font-family` and `font-size` of its parent or `<body>` element. This rule ensures the semantics of a heading element are distinct from its appearance. If you need to style a heading, refer to our [headings](/foundation/typography/headings/) or [utility class](/utilities/overview/) documentation.

Reviewer: > This rule ensures the semantics of a heading element are distinct from its appearance. If you need to style a heading, refer to our [headings](/foundation/typography/headings/) or [utility class](/utilities/overview/) documentation.

💯

> Lastly, headings have been reset to have their bottom margins removed. All headings have a top margin set to `1.5em`.

Reviewer: Should we explain why it's `1.5` instead of on the 8px grid? Because it's relative to the heading's font size?

Author: I'm not sure if we need to explain why it's `1.5em`... part of me wonders if this line should be removed entirely. I do specify on line 22 that we're pivoting to using top margins only and maybe that's enough? By hard-coding a value here, does this run the risk of falling out of date with the codebase as we go? And does this margin apply to all themes? I think including this info raises more questions than it solves.

Reviewer: Yeah, that's a good point. I think it's nice to at least call out that it has top margins and not bottom margins

### Paragraphs

> Like headings, paragraphs have had their bottom margins removed and a top margin of `1em` defined. However, if a paragraph is the `:first-child` its margin will be entirely removed.

Reviewer: > However, if a paragraph is the `:first-child` its margin will be entirely removed.

Should we give the same treatment to headings? Has that been talked about?

Author: That's a good thought. Our heading style classes currently handle that behavior, but we could easily lift that to the HTML elements

Author: Should I update the reset to include the `:first-child`/`:last-child` rules from our [headings CSS](https://github.com/CMSgov/design-system/blob/main/packages/design-system/src/styles/_base.scss#L158)?

Reviewer: Oh, I didn't know we were doing that! Yeah, I think moving that into the reset is a good idea, especially if the values don't change per heading level.

### Links

> Links behave as they previously did, with one exception - the `text-decoration-color` is now set to `currentColor`. This change was made under the assumption that all link underlines will have the same color as its text, regardless of that text's state.

Reviewer: ```suggestion
Links behave as they previously did, with one exception—the `text-decoration-color` is now set to `currentColor`. This change was made under the assumption that all link underlines will have the same color as its text, regardless of that text's state.
```

Emdash with no spaces is what you want here, I believe.

I'm also inclined to use a colon for this sort of thing. I've gotten more liberal with the `:` over the years. Some may say that the semi-colon is more appropriate, but I don't think most readers expect it in this sort of situation.

## Lists

The `<ol>` and `<ul>` elements have been styled to match their [`.ds-c-list`](/foundation/typography/list/) counterpart, with one key difference - they use `flex` to better control list item spacing and assist with commonly used inline list patterns.

If you find yourself in need of an inline list, you can add `flex-direction: row` locally to your project. Likewise, if the spacing between list elements isn't to your liking, `gap` can be changed to the value that fits your project best.

> One last change worth mentioning for lists is that unstyled lists are defined under the `role="list"` attribute. This makes something like `<ul role="list">` visually the same as applying the `.ds-c-list--bare` class. Because unstyled lists have [accessibility concerns](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style-type#accessibility_concerns), we opted to include this style rule to ensure proper markup is used.

Reviewer: Love it

## Form elements

Various form elements have been rebooted for simpler base styles. Here are some of the most notable changes:

- The `<input>`, `<button>`, `<textarea>`, and `<select>` elements receive the font styles defined by their parent element using the `font: inherit` rule. 

> - All `:disabled` fields have `cursor: not-allowed` to prevent user confusion.

Reviewer: Did you talk to Dan about this? Or did this stem from some other source?

Author: This was something Eric Bailey and I discussed during the HCM work. Mainly, disabled states typically rely on color alone to indicate what it is and adding the cursor was just an extra indicator for users. But we can get confirmation from Dan just in case.
